Nightly builds run out of memory

I have just noticed that the nightly build action in [solc-bin](https://github.com/ethereum/solc-bin) has been failing for three days in a row now: [1](https://github.com/ethereum/solc-bin/runs/1201552073?check_suite_focus=true), [2](https://github.com/ethereum/solc-bin/runs/1204181256?check_suite_focus=true), [3](https://github.com/ethereum/solc-bin/runs/1206816217?check_suite_focus=true).

The output suggests that it's being killed by github.

I suspect it's either running out of memory or out of time and the former seems more likely. I just ran `./update` locally and RAM usage peaks at 11.4 GB. It's somewhere around 10 GB right about the point where it fails on github.

### The error
```
Computing hashes of '/bin/soljson-v0.7.3-nightly.2020.9.28+commit.dd5b0a71.js'
Computing hashes of '/bin/soljson-v0.7.3-nightly.2020.9.29+commit.343c13f9.js'
Computing hashes of '/bin/soljson-v0.7.3-nightly.2020.9.30+commit.3af21c92.js'
Killed
npm ERR! code ELIFECYCLE
npm ERR! errno 137
npm ERR! solc-bin@0.0.0 update: `./update`
npm ERR! Exit status 137
npm ERR! 
npm ERR! Failed at the solc-bin@0.0.0 update script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2020-10-05T01_05_31_133Z-debug.log
Error: Process completed with exit code 137.
```

The problem came back: https://github.com/ethereum/solc-bin/runs/4438376804
```
Computing hashes of '/bin/soljson-v0.4.16-nightly.2017.8.10+commit.41e3cbe0.js'
Computing hashes of '/bin/soljson-v0.4.15-nightly.2017.8.1+commit.7e07eb6e.js'
Computing hashes of '/windows-amd64/solc-windows-amd64-v0.4.17+commit.bdeb9e52.zip'
Computing hashes of '/windows-amd64/solc-windows-amd64-v0.4.18+commit.9cf6e910.zip'
Computing hashes of '/windows-amd64/solc-windows-amd64-v0.4.19+commit.c4cbbb05.zip'
Computing hashes of '/windows-amd64/solc-windows-amd64-v0.5.5+commit.47a71e8f.zip'
Killed
Error: Process completed with exit code 137.
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -5,7 +5,7 @@ jobs:
     docker:
       # FIXME: sha3 package fails to build on latest node.js (version 17)
       - image: circleci/node:16
-    resource_class: xlarge
+    resource_class: small
     steps:
       - checkout
       - run: npm install

--- a/package.json
+++ b/package.json
@@ -16,7 +16,7 @@
   },
   "scripts": {
     "lint": "standard update",
-    "update": "./update",
+    "update": "./update --max-files-per-batch 1",
     "test": "git checkout -f && npm run update && git diff --exit-code"
   },
   "repository": {

--- a/update
+++ b/update
@@ -155,6 +155,18 @@ async function makeEntry (dir, parsedFileName, oldList) {
   return build
 }
 
+async function batchedAsyncMap (values, batchSize, asyncMapFunction) {
+  if (batchSize === null) {
+    batchSize = values.length
+  }
+
+  let results = []
+  for (let i = 0; i < values.length; i += batchSize) {
+    results = results.concat(await Promise.all(values.slice(i, i + batchSize).map(asyncMapFunction)))
+  }
+  return results
+}
+
 function processDir (dir, options, listCallback) {
   fs.readdir(path.join(__dirname, dir), { withFileTypes: true }, async function (err, files) {
     if (err) {
@@ -183,25 +195,26 @@ function processDir (dir, options, listCallback) {
     }[dir] || ''
 
     // ascending list (oldest version first)
-    const parsedList = (await Promise.all(
-      files
-        .filter(function (file) {
-          // Skip symbolic links with less then 8 characters in the commit hash.
-          // They exist only for backwards-compatibilty and should not be on the list.
-          return dir !== '/bin' ||
-            !file.isSymbolicLink() ||
-            file.name.match(/^.+\+commit\.[0-9a-f]{8,}\.js$/)
+    const parsedFileNames = files
+      .filter(function (file) {
+        // Skip symbolic links with less then 8 characters in the commit hash.
+        // They exist only for backwards-compatibilty and should not be on the list.
+        return dir !== '/bin' ||
+          !file.isSymbolicLink() ||
+          file.name.match(/^.+\+commit\.[0-9a-f]{8,}\.js$/)
+      })
+      .map(function (file) { return file.name })
+      .map(function (binaryName) {
+        const escapedExtensions = binaryExtensions.map(function (binaryExtension) {
+          return binaryExtension.replace('.', '\\.')
         })
-        .map(function (file) { return file.name })
-        .map(function (binaryName) {
-          const escapedExtensions = binaryExtensions.map(function (binaryExtension) {
-            return binaryExtension.replace('.', '\\.')
-          })
-          return binaryName.match(new RegExp('^' + binaryPrefix + '-v([0-9.]*)(-([^+]*))?(\\+(.*))?(' + escapedExtensions.join('|') + ')$'))
-        })
-        .filter(function (matchResult) { return matchResult !== null })
-        .map(async function (matchResult) { return await makeEntry(dir, matchResult, oldList) })
-    ))
+        return binaryName.match(new RegExp('^' + binaryPrefix + '-v([0-9.]*)(-([^+]*))?(\\+(.*))?(' + escapedExtensions.join('|') + ')$'))
+      })
+      .filter(function (matchResult) { return matchResult !== null })
+
+    const parsedList = (await batchedAsyncMap(parsedFileNames, options.maxFilesPerBatch, async function (matchResult) {
+      return await makeEntry(dir, matchResult, oldList)
+    }))
       .sort(function (a, b) {
         if (a.longVersion === b.longVersion) {
           return 0
@@ -317,10 +330,23 @@ function processDir (dir, options, listCallback) {
 
 function parseCommandLine () {
   let reuseHashes
+  let maxFilesPerBatch
 
   for (let i = 2; i < process.argv.length; ++i) {
     if (process.argv[i] === '--reuse-hashes') {
       reuseHashes = true
+    } else if (process.argv[i] === '--max-files-per-batch') {
+      if (i + 1 >= process.argv.length) {
+        console.error('Expected an integer argument after --max-files-per-batch.')
+        process.exit(1)
+      }
+
+      maxFilesPerBatch = parseInt(process.argv[i + 1], 10)
+      if (isNaN(maxFilesPerBatch) || maxFilesPerBatch <= 0) {
+        console.error("Expected the argument of --max-files-per-batch to be a positive integer, got '" + process.argv[i + 1] + "'.")
+        process.exit(1)
+      }
+      ++i
     } else {
       console.error("Invalid option: '" + process.argv[i] + "'.")
       process.exit(1)
@@ -331,9 +357,13 @@ function parseCommandLine () {
   if (reuseHashes === undefined) {
     reuseHashes = false
   }
+  if (maxFilesPerBatch === undefined) {
+    maxFilesPerBatch = null // no limit
+  }
 
   return {
     reuseHashes: reuseHashes,
+    maxFilesPerBatch: maxFilesPerBatch
   }
 }
 

--- a/update
+++ b/update
@@ -316,18 +316,24 @@ function processDir (dir, options, listCallback) {
 }
 
 function parseCommandLine () {
-  if (process.argv.length > 3) {
-    console.error('Expected at most one argument, got ' + (process.argv.length - 2))
-    process.exit(1)
+  let reuseHashes
+
+  for (let i = 2; i < process.argv.length; ++i) {
+    if (process.argv[i] === '--reuse-hashes') {
+      reuseHashes = true
+    } else {
+      console.error("Invalid option: '" + process.argv[i] + "'.")
+      process.exit(1)
+    }
   }
 
-  if (process.argv.length === 3 && process.argv[2] !== '--reuse-hashes') {
-    console.error('Invalid argument: ' + process.argv[2])
-    process.exit(1)
+  // Defaults
+  if (reuseHashes === undefined) {
+    reuseHashes = false
   }
 
   return {
-    reuseHashes: process.argv.length === 3 && process.argv[2] === '--reuse-hashes'
+    reuseHashes: reuseHashes,
   }
 }
 

--- a/update
+++ b/update
@@ -183,21 +183,24 @@ function processDir (dir, options, listCallback) {
     }[dir] || ''
 
     // ascending list (oldest version first)
-    const parsedList = (await Promise.all(files
-      .filter(function (file) {
-        // Skip symbolic links with less then 8 characters in the commit hash.
-        // They exist only for backwards-compatibilty and should not be on the list.
-        return dir !== '/bin' ||
-          !file.isSymbolicLink() ||
-          file.name.match(/^.+\+commit\.[0-9a-f]{8,}\.js$/)
-      })
-      .map(function (file) { return file.name })
-      .map(function (file) {
-        const escapedExtensions = binaryExtensions.map(function (binaryExtension) { return binaryExtension.replace('.', '\\.') })
-        return file.match(new RegExp('^' + binaryPrefix + '-v([0-9.]*)(-([^+]*))?(\\+(.*))?(' + escapedExtensions.join('|') + ')$'))
-      })
-      .filter(function (version) { return version })
-      .map(async function (pars) { return await makeEntry(dir, pars, oldList) })
+    const parsedList = (await Promise.all(
+      files
+        .filter(function (file) {
+          // Skip symbolic links with less then 8 characters in the commit hash.
+          // They exist only for backwards-compatibilty and should not be on the list.
+          return dir !== '/bin' ||
+            !file.isSymbolicLink() ||
+            file.name.match(/^.+\+commit\.[0-9a-f]{8,}\.js$/)
+        })
+        .map(function (file) { return file.name })
+        .map(function (binaryName) {
+          const escapedExtensions = binaryExtensions.map(function (binaryExtension) {
+            return binaryExtension.replace('.', '\\.')
+          })
+          return binaryName.match(new RegExp('^' + binaryPrefix + '-v([0-9.]*)(-([^+]*))?(\\+(.*))?(' + escapedExtensions.join('|') + ')$'))
+        })
+        .filter(function (matchResult) { return matchResult !== null })
+        .map(async function (matchResult) { return await makeEntry(dir, matchResult, oldList) })
     ))
       .sort(function (a, b) {
         if (a.longVersion === b.longVersion) {
@@ -217,7 +220,8 @@ function processDir (dir, options, listCallback) {
 
     // descending list
     const releases = parsedList
-      .slice().reverse()
+      .slice()
+      .reverse()
       .reduce(function (prev, next) {
         if (next.prerelease === undefined) {
           prev[next.version] = next.path
@@ -227,19 +231,21 @@ function processDir (dir, options, listCallback) {
 
     // descending list
     const buildNames = parsedList
-      .slice().reverse()
-      .map(function (ver) { return ver.path })
+      .slice()
+      .reverse()
+      .map(function (listEntry) { return listEntry.path })
 
     const latestRelease = parsedList
-      .slice().reverse()
-      .filter(function (version) {
-        if (version.prerelease === undefined) {
-          return version
+      .slice()
+      .reverse()
+      .filter(function (listEntry) {
+        if (listEntry.prerelease === undefined) {
+          return listEntry
         }
         return undefined
       })
-      .map(function (version) {
-        return version.version
+      .map(function (listEntry) {
+        return listEntry.version
       })[0]
 
     // latest build (nightly)

